File: taxcalcbench/gen_correct.ts
#!/usr/bin/env -S deno run --allow-read --allow-write
/**
 * Generate correct.json for each benchmark case using the 2025 tax calculator.
 * Run: deno run --allow-read --allow-write taxcalcbench/gen_correct.ts
 */

import { dirname, fromFileUrl, join } from "@std/path";
import { computeTax, type FilingStatus, type TaxReturnInput } from "./tax2025.ts";

const CASES_DIR = join(dirname(fromFileUrl(import.meta.url)), "cases");

interface Dependent { qualifying_child_for_ctc?: boolean; }
interface FormEntry { node_type: string; data: Record<string, unknown>; }
interface CaseInput { year: number; scenario?: string; forms: FormEntry[]; }

function buildReturn(caseData: CaseInput): TaxReturnInput {
  const forms = caseData.forms;
  let general: Record<string, unknown> = {};
  for (const f of forms) {
    if (f.node_type === "start") { general = (f.data.general ?? {}) as Record<string, unknown>; break; }
  }

  const status      = (general.filing_status ?? "single") as FilingStatus;
  const dependents  = (general.dependents ?? []) as Dependent[];

  const inp: TaxReturnInput = {
    status,
    taxpayer65:    (general.taxpayer_age_65_or_older ?? false) as boolean,
    taxpayerBlind: (general.taxpayer_blind           ?? false) as boolean,
    spouse65:      (general.spouse_age_65_or_older   ?? false) as boolean,
    spouseBlind:   (general.spouse_blind             ?? false) as boolean,
    qualifyingChildren: dependents.filter(d => d.qualifying_child_for_ctc).length,
    eitcChildren:       dependents.length,
    wages: 0, unemployment: 0, interest: 0, ordinaryDividends: 0, qualifiedDividends: 0,
    ltcg: 0, pension: 0, ssaGross: 0, scheduleCNet: 0,
    studentLoanInterestPaid: 0, hsaEmployer: 0,
    educatorExpenses: 0, educatorExpensesSp: 0,
    depCareExpenses: 0, depCarePersons: 0, aotcExpenses: 0,
    marketplacePremium: 0, marketplaceSlcsp: 0, marketplaceAptc: 0,
    fedWithheld: 0, estimatedTaxPayments: 0,
    ssWagesList: [], ssWithheldList: [],
  };

  for (const f of forms) {
    const nt = f.node_type;
    const d  = f.data as Record<string, number>;

    if (nt === "w2") {
      inp.wages!         += d.box1_wages        ?? 0;
      inp.fedWithheld!   += d.box2_fed_withheld ?? 0;
      inp.ssWagesList!.push(d.box3_ss_wages    ?? 0);
      inp.ssWithheldList!.push(d.box4_ss_withheld ?? 0);
      for (const e of (f.data.box12_entries ?? []) as Array<{code:string;amount:number}>) {
        if (e.code === "W") inp.hsaEmployer! += e.amount ?? 0;
      }
    } else if (nt === "f1099g") {
      inp.unemployment!  += d.box_1_unemployment        ?? 0;
      inp.fedWithheld!   += d.box_4_federal_withheld    ?? 0;
    } else if (nt === "f1099int") {
      inp.interest!      += d.box1 ?? 0;
      inp.fedWithheld!   += d.box4 ?? 0;
    } else if (nt === "f1099div") {
      inp.ordinaryDividends!  += d.box1a ?? 0;
      inp.qualifiedDividends! += d.box1b ?? 0;
      inp.fedWithheld!        += d.box4  ?? 0;
    } else if (nt === "f1099b") {
      const gain = (d.proceeds ?? 0) - (d.cost_basis ?? 0);
      const part = ((f.data.part ?? "") as string).toUpperCase();
      if (part === "D" || part === "") inp.ltcg! += gain;
    } else if (nt === "f1099r") {
      inp.pension!       += d.box2a_taxable_amount ?? d.box1_gross_distribution ?? 0;
      inp.fedWithheld!   += d.box4_federal_withheld ?? 0;
    } else if (nt === "ssa1099") {
      inp.ssaGross!      += d.box3_gross_benefits    ?? 0;
      inp.fedWithheld!   += d.box6_federal_withheld  ?? 0;
    } else if (nt === "schedule_c") {
      const gross = d.line_1_gross_receipts ?? 0;
      const skip  = new Set(["line_1_gross_receipts","line_a_principal_business",
                             "line_b_business_code","line_f_accounting_method","line_g_material_participation"]);
      let expenses = 0;
      for (const [k, v] of Object.entries(d)) {
        if (k.startsWith("line_") && !skip.has(k) && typeof v === "number") expenses += v;
      }
      inp.scheduleCNet! += gross - expenses;
    } else if (nt === "f1099nec") {
      inp.scheduleCNet! += d.box1_nec            ?? 0;
      inp.fedWithheld!  += d.box4_federal_withheld ?? 0;
    } else if (nt === "educator_expenses") {
      inp.educatorExpenses!   += d.educator1_expenses ?? 0;
      inp.educatorExpensesSp! += d.educator2_expenses ?? 0;
    } else if (nt === "f1040es") {
      inp.estimatedTaxPayments! += (d.payment_q1 ?? 0) + (d.payment_q2 ?? 0) +
                                   (d.payment_q3 ?? 0) + (d.payment_q4 ?? 0);
    } else if (nt === "f8863") {
      let entries: Array<Record<string,unknown>>;
      if (f.data.f8863s)       entries = f.data.f8863s as Array<Record<string,unknown>>;
      else if (f.data.credit_type) entries = [f.data];
      else entries = [];
      for (const e of entries) {
        if (e.credit_type === "aoc") inp.aotcExpenses! += (e.aoc_adjusted_expenses as number) ?? 0;
      }
    } else if (nt === "f1095a") {
      let plans: Array<Record<string,unknown>>;
      if (f.data.f1095as)      plans = f.data.f1095as as Array<Record<string,unknown>>;
      else if (f.data.issuer_name) plans = [f.data];
      else plans = [];
      for (const p of plans) {
        inp.marketplacePremium! += (p.annual_premium as number) ?? 0;
        inp.marketplaceSlcsp!   += (p.annual_slcsp   as number) ?? 0;
        inp.marketplaceAptc!    += (p.annual_aptc    as number) ?? 0;
      }
    } else if (nt === "f1098e") {
      // Input format: flat fields at root (no f1098es wrapper)
      if (typeof f.data.box1_student_loan_interest === "number") {
        inp.studentLoanInterestPaid! += f.data.box1_student_loan_interest;
      } else {
        // Fallback: wrapped array format
        for (const e of (f.data.f1098es ?? []) as Array<{box1_student_loan_interest?:number}>) {
          inp.studentLoanInterestPaid! += e.box1_student_loan_interest ?? 0;
        }
      }
    } else if (nt === "f2441") {
      inp.depCareExpenses = d.qualifying_expenses_paid ?? 0;
      inp.depCarePersons  = d.qualifying_person_count  ?? 1;
    }
  }

  return inp;
}

async function main() {
  console.log("Generating correct.json for all cases...\n");

  const names: string[] = [];
  for await (const e of Deno.readDir(CASES_DIR)) {
    if (e.isDirectory) names.push(e.name);
  }
  names.sort();

  for (const name of names) {
    const caseDir    = join(CASES_DIR, name);
    const inputFile  = join(caseDir, "input.json");
    const outputFile = join(caseDir, "correct.json");
    try { await Deno.stat(inputFile); } catch { continue; }

    const caseData: CaseInput = JSON.parse(await Deno.readTextFile(inputFile));
    const inp    = buildReturn(caseData);
    const result = computeTax(inp);

    await Deno.writeTextFile(outputFile, JSON.stringify({
      case:     name,
      scenario: caseData.scenario ?? "",
      year:     2025,
      inputs: {
        filing_status:           inp.status,
        wages:                   inp.wages,
        unemployment:            inp.unemployment,
        interest:                inp.interest,
        ordinary_dividends:      inp.ordinaryDividends,
        qualified_dividends:     inp.qualifiedDividends,
        ltcg:                    inp.ltcg,
        pension:                 inp.pension,
        ssa_gross:               inp.ssaGross,
        schedule_c_net:          inp.scheduleCNet,
        student_loan_int:        inp.studentLoanInterestPaid,
        hsa_employer:            inp.hsaEmployer,
        educator_expenses:       inp.educatorExpenses,
        educator_expenses_sp:    inp.educatorExpensesSp,
        aotc_expenses:           inp.aotcExpenses,
        marketplace_premium:     inp.marketplacePremium,
        marketplace_slcsp:       inp.marketplaceSlcsp,
        marketplace_aptc:        inp.marketplaceAptc,
        qualifying_children:     inp.qualifyingChildren,
        eitc_children:           inp.eitcChildren,
        dep_care_expenses:       inp.depCareExpenses,
        estimated_tax_payments:  inp.estimatedTaxPayments,
        fed_withheld:            inp.fedWithheld,
      },
      correct: result,
    }, null, 2));

    const r = result;
    const n = (v: number) => v.toLocaleString("en-US", { minimumFractionDigits: 2 });
    console.log(`  ${name}: agi=${n(r.line11_agi).padStart(12)}  tax=${n(r.line24_total_tax).padStart(10)}  refund=${n(r.line35a_refund).padStart(10)}  owed=${n(r.line37_amount_owed).padStart(9)}`);
  }

  console.log("\nDone.");
}

main();

File: taxcalcbench/gen_correct.py
#!/usr/bin/env python3
"""
Generate correct.json for each benchmark case using the 2025 tax calculator.
Run from the benchmark/ directory: python3 gen_correct.py
"""
import json
import sys
from pathlib import Path
from tax2025 import TaxReturn

CASES_DIR = Path(__file__).parent / "cases"

def build_return(case_data: dict) -> TaxReturn:
    forms = case_data["forms"]
    general = {}
    for f in forms:
        if f["node_type"] == "start":
            general = f["data"].get("general", {})
            break

    status = general.get("filing_status", "single")
    t65    = general.get("taxpayer_age_65_or_older", False)
    t_blind = general.get("taxpayer_blind", False)
    s65    = general.get("spouse_age_65_or_older", False)
    s_blind = general.get("spouse_blind", False)
    dependents = general.get("dependents", [])

    qualifying_children = sum(1 for d in dependents if d.get("qualifying_child_for_ctc"))
    eitc_children = len(dependents)  # simplification: all dependents count for EITC

    r = TaxReturn(
        status=status,
        taxpayer_65=t65,
        taxpayer_blind=t_blind,
        spouse_65=s65,
        spouse_blind=s_blind,
        qualifying_children=qualifying_children,
        eitc_children=eitc_children,
    )

    ss_wages_list = []
    ss_withheld_list = []

    for f in forms:
        nt = f["node_type"]
        d  = f["data"]

        if nt == "w2":
            r.wages        += d.get("box1_wages", 0)
            r.fed_withheld += d.get("box2_fed_withheld", 0)
            ss_wages_list.append(d.get("box3_ss_wages", 0))
            ss_withheld_list.append(d.get("box4_ss_withheld", 0))
            for entry in d.get("box12_entries", []):
                if entry.get("code") == "W":
                    r.hsa_employer += entry.get("amount", 0)

        elif nt == "f1099g":
            r.unemployment += d.get("box_1_unemployment", 0)
            r.fed_withheld += d.get("box_4_federal_withheld", 0)

        elif nt == "f1099int":
            r.interest     += d.get("box1", 0)
            r.fed_withheld += d.get("box4", 0)

        elif nt == "f1099div":
            r.ordinary_dividends  += d.get("box1a", 0)
            r.qualified_dividends += d.get("box1b", 0)
            r.fed_withheld        += d.get("box4", 0)

        elif nt == "f1099b":
            proceeds = d.get("proceeds", 0)
            basis    = d.get("cost_basis", 0)
            # Only long-term (Part D = long-term)
            part = d.get("part", "")
            gain = proceeds - basis
            if part.upper() in ("D", ""):
                r.ltcg += gain

        elif nt == "f1099r":
            r.pension      += d.get("box2a_taxable_amount", d.get("box1_gross_distribution", 0))
            r.fed_withheld += d.get("box4_federal_withheld", 0)

        elif nt == "ssa1099":
            r.ssa_gross += d.get("box3_gross_benefits", 0)
            r.fed_withheld += d.get("box6_federal_withheld", 0)

        elif nt == "schedule_c":
            gross    = d.get("line_1_gross_receipts", 0)
            expenses = sum(v for k, v in d.items()
                           if k.startswith("line_") and k not in
                           ("line_1_gross_receipts", "line_a_principal_business",
                            "line_b_business_code", "line_f_accounting_method",
                            "line_g_material_participation")
                           and isinstance(v, (int, float)))
            r.schedule_c_net += gross - expenses

        elif nt == "f1099nec":
            r.schedule_c_net += d.get("box1_nec", 0)
            r.fed_withheld   += d.get("box4_federal_withheld", 0)

        elif nt == "educator_expenses":
            r.educator_expenses    += d.get("educator1_expenses", 0)
            r.educator_expenses_sp += d.get("educator2_expenses", 0)

        elif nt == "f1040es":
            r.estimated_tax_payments += sum(
                d.get(k, 0) for k in ["payment_q1", "payment_q2", "payment_q3", "payment_q4"]
            )

        elif nt == "f8863":
            # Support both flat format (single item at root) and wrapped format
            if "f8863s" in d:
                entries = d["f8863s"]
            elif "credit_type" in d:
                entries = [d]  # flat format — single student entry
            else:
                entries = []
            for s_entry in entries:
                if s_entry.get("credit_type") == "aoc":
                    r.aotc_expenses += s_entry.get("aoc_adjusted_expenses", 0)

        elif nt == "f1095a":
            # Support both flat format (single item at root) and wrapped format
            if "f1095as" in d:
                plans = d["f1095as"]
            elif "issuer_name" in d:
                plans = [d]  # flat format — single policy entry
            else:
                plans = []
            for plan in plans:
                r.marketplace_premium += plan.get("annual_premium", 0)
                r.marketplace_slcsp   += plan.get("annual_slcsp", 0)
                r.marketplace_aptc    += plan.get("annual_aptc", 0)

        elif nt == "f1098e":
            for entry in d.get("f1098es", []):
                r.student_loan_interest_paid += entry.get("box1_student_loan_interest", 0)

        elif nt == "f2441":
            r.dep_care_expenses = d.get("qualifying_expenses_paid", 0)
            r.dep_care_persons  = d.get("qualifying_person_count", 1)

    r.ss_wages_list    = ss_wages_list
    r.ss_withheld_list = ss_withheld_list
    return r


def main():
    cases = sorted(CASES_DIR.iterdir())
    for case_dir in cases:
        if not case_dir.is_dir():
            continue
        input_file   = case_dir / "input.json"
        correct_file = case_dir / "correct.json"
        if not input_file.exists():
            continue

        with open(input_file) as f:
            case_data = json.load(f)

        tr = build_return(case_data)
        result = tr.compute()

        output = {
            "case":     case_dir.name,
            "scenario": case_data.get("scenario", ""),
            "year":     2025,
            "inputs": {
                "filing_status":     tr.status,
                "wages":             tr.wages,
                "unemployment":      tr.unemployment,
                "interest":          tr.interest,
                "ordinary_dividends": tr.ordinary_dividends,
                "qualified_dividends": tr.qualified_dividends,
                "ltcg":              tr.ltcg,
                "pension":           tr.pension,
                "ssa_gross":         tr.ssa_gross,
                "schedule_c_net":    tr.schedule_c_net,
                "student_loan_int":  tr.student_loan_interest_paid,
                "hsa_employer":      tr.hsa_employer,
                "educator_expenses": tr.educator_expenses,
                "educator_expenses_sp": tr.educator_expenses_sp,
                "aotc_expenses":     tr.aotc_expenses,
                "marketplace_premium": tr.marketplace_premium,
                "marketplace_slcsp": tr.marketplace_slcsp,
                "marketplace_aptc":  tr.marketplace_aptc,
                "qualifying_children": tr.qualifying_children,
                "eitc_children":     tr.eitc_children,
                "dep_care_expenses": tr.dep_care_expenses,
                "estimated_tax_payments": tr.estimated_tax_payments,
                "fed_withheld":      tr.fed_withheld,
            },
            "correct": result,
        }

        with open(correct_file, "w") as f:
            json.dump(output, f, indent=2)
        print(f"  {case_dir.name}: agi={result['line11_agi']:>10,.2f}  "
              f"tax={result['line24_total_tax']:>9,.2f}  "
              f"refund={result['line35a_refund']:>9,.2f}  "
              f"owed={result['line37_amount_owed']:>8,.2f}")

if __name__ == "__main__":
    print("Generating correct.json for all cases...\n")
    main()
    print("\nDone.")
